### PR TITLE
feat: New workflow action to send Whatsapp message

### LINKED ISSUE #3664: [CAL-425] New workflow action to send Whatsapp message
### Is your proposal related to a problem?

No.

### Describe the solution you'd like

I want to send confirmation mails with video links to attendees over whatsapp, instead of their mails.

### Describe alternatives you've considered

[Calenderly.com](http://Calenderly.com) has workflows that allow me to do this.

### Additional context

(Write your answer here.)

> From [SyncLinear.com](https://synclinear.com)

### PR DESCRIPTION
**Pull request by jemiluv8**
## What does this PR do?
Add Support for the following whatsapp workflows

- send Whatsapp message to attendee
- send Whatsapp message to specific number

### Loom Demos
[Creating a whatsapp workflow by editing existing actions](https://www.loom.com/share/baa9de91fb40401d9d04d888d22c9a78)

[Creating whatsapp workflow action using the AddActionDialog](https://www.loom.com/share/2af273aaac7d4d9a821a0bdf6e82cdd1)

[Book meeting that has a whatsapp workflow to send whatsapp to attendee](https://www.loom.com/share/576844046e024772b6fc20dee9bce40c)

**Environment**: Staging(main branch) / Production
Fixes #3664
/claim #3664
## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How should this be tested?

- [ ] Get a sandbox whatsapp twilio number for the environment variable - TWILIO_WHATSAPP_PHONE_NUMBER
- [ ] Send a whatsapp message with the content `join liquid-rubber` to the twilio whatsapp number obtained
- [ ] Create a workflow to send whatsapp message to attendee or a specific number
- [ ] Verify that a whatsapp message is sent to your number 

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by jemiluv8** [comment]
@CarinaWolli, I just created a draft PR to get some feedback on next steps.
There are a few things to “sort” out.

### WhatsApp Platform Issues

WhatsApp allows sending messages using only [approved templates](https://support.twilio.com/hc/en-us/articles/360007721954-Getting-Started-with-Twilio-for-WhatsApp). These templates can be managed via the Twilio account. This means we can’t have “CUSTOM” templates for WhatsApp workflow actions.

This limitation led me to create a couple of “default/proposed” templates for “WhatsApp” actions based on the event “triggers”. So there are templates for NEW_EVENT, EVENT_CANCELLED and the likes. Based on the platform limitations this made sense to me.

I expect we’d only need a handful of useful templates for sending things like meeting links.
#### I need assistance to decide on the templates we need and their content.

### WhatsApp Templates and UX

- I don’t allow editing any of the WhatsApp templates. This is supposed to be readonly since we can only send WhatsApp messages using “approved” templates.

**Comment by CarinaWolli** [comment]
@jemiluv8 awesome work so far 🙌🏻

> I need assistance to decide on the templates we need and their content. 

@Jaibles could you take a look at that? Here are the templates that got defined in that PR:
Screenshot 2023-05-10 at 14 56 39
Screenshot 2023-05-10 at 14 56 43
Screenshot 2023-05-10 at 14 56 47

**Comment by jemiluv8** [comment]
@Jaibles is there any update on the content for the Whatsapp templates?

**Comment by Jaibles** [comment]
> @Jaibles is there any update on the content for the Whatsapp templates?

Sorry, I missed this. I will take a look now.

**Comment by Jaibles** [comment]
@jemiluv8 Can we add any formatting? Some bolds for the event titles would be good. 

Some small changes

**Cancelled**
Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on
{EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE} has been cancelled.

**Completed**
Hi {ATTENDEE}, thank you for attending the event ({EVENT_NAME}) on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE}.

**Rescheduled**
Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on
{EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE} has been rescheduled.

**Comment by jemiluv8** [comment]
> @jemiluv8 Can we add any formatting? Some bolds for the event titles would be good.
> 
> Some small changes
> 
> **Cancelled** Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}) has been cancelled.
> 
> **Completed** Hi {ATTENDEE}, thank you for attending the event ({EVENT_NAME}) on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}).
> 
> **Rescheduled** Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}) has been rescheduled.

@Jaibles that is great. I'll bolden the event titles.

File: packages/features/bookings/lib/getBookingResponsesSchema.ts
Comment (jemiluv8) [inline packages/features/bookings/lib/getBookingResponsesSchema.ts:181]: @CarinaWolli,

I had the issue in the screenshot below and ended up having to "patch" this file and the zod schema validation.
I don't like my patch and have been investigating the source of the missing "+" from phone number inputs.

A recent commit however changed the react library used for our phone input from "react-phone-number-input" 
to "react-phone-input-2". I'm not sure what to make of this yet but prior to that the "+" prefix was added to the
value of the phone number input.

I'm reaching out to decide what to do about this. I wonder if bookings that require entering phone number are going through (assuming the recent changes have been deployed)

Screenshot 2023-06-01 at 12 25 53 PM

Comment (CarinaWolli) [reply packages/features/bookings/lib/getBookingResponsesSchema.ts:181]: This has been fixed already here: https://github.com/calcom/cal.com/pull/9258

Sorry for that!

Comment (jemilu2) [reply packages/features/bookings/lib/getBookingResponsesSchema.ts:181]: Thanks.

**Comment by PeerRich** [comment]
thank you for marking as "for review". we will try to review this this week, there are some other higher priority PRs right now tho which have to be reviewed first: https://github.com/calcom/cal.com/pulls?q=is:pr+is:open+sort:updated-desc+label:%22High+priority%22

File: packages/features/ee/workflows/lib/actionHelperFunctions.ts
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/actionHelperFunctions.ts:74]: new line missing

File: packages/features/ee/workflows/lib/constants.ts
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/constants.ts:37]: We can also add these new templates for SMS (but we can do that in a follow-up PR)

File: packages/features/ee/workflows/lib/reminders/templates/whatsapp/index.ts
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/reminders/templates/whatsapp/index.ts:4]: new line missing

File: packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql
Comment (CarinaWolli) [inline packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql:6]: @zomars are we supporting PostresSQL version 11 and earlier? If yes, we would need a separate migration file for each enum value added

**Review by CarinaWolli — COMMENTED** [review]
Great work on what I have seen so far 👏🏻  I left you some first feedback, but I didn't find the time yet to review and test in detail. I also started to look into what needs to be done to make it work in production and submitted one first template to Twilio which is waiting for approval currently :) 

Maybe you can already address my comments, and I will come back soon to finish reviewing

File: packages/features/ee/workflows/components/WorkflowStepContainer.tsx
Comment (CarinaWolli) [inline packages/features/ee/workflows/components/WorkflowStepContainer.tsx:773]: type error
Screenshot 2023-06-15 at 10 44 43

File: packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts
Comment (jemiluv8) [reply packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts:31]: No I don't know something different. I was actually using getDefaultSender for whatsapp initially. Will change it back.

Comment (jemiluv8) [reply packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts:31]: I have thus far relied on the implementation of SMS workflow as a guide. And when sending to attendee the SMS workflow shows a SENDER id similar to what I do for Whatsapp workflow(in the screenshot below).

As far as I was concerned any valid twilio whatsapp number can be used as a Sender Id (just like it is for SMS). The current implementation uses the Sender ID when  specified. The change you suggested will just ignore it.

To be consistent with everything else, I will also need to remove the Sender ID input box in the screenshot below for whatsapp related workflows. 

Should I proceed with both changes?

Screenshot 2023-06-15 at 9 55 56 AM

File: packages/features/ee/workflows/lib/constants.ts
Comment (jemiluv8) [reply packages/features/ee/workflows/lib/constants.ts:37]: I'll be happy to do it in a follow-up PR. This one feels too big already.

File: packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts
Comment (CarinaWolli) [reply packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts:31]: Yes, please make both of the changes. Thank you 🙏

Comment (jemiluv8) [reply packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts:31]: I just made both changes.

**Comment by CarinaWolli** [comment]
I am currently working on getting the WhatsApp templates approved in Twilio. I created the reminder template at first and it got approved right away. But canceled and rescheduled template got rejected, without any reason given. I resubmitted it and will see if it goes through this time

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
I tested it and finished my review. It all looks good to me, I just left you some minor comments. Please address them and I will do my best to get the Twilio templates approved, so we can merge it. Thank you 🙏

File: packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql
Comment (CarinaWolli) [reply packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql:6]: Should be fine to have them in one migration file 👍
Could you add all migrations into one single migration file?

File: packages/features/ee/workflows/lib/reminders/templates/whatsapp/whatsappEventCancelledTemplate.ts
Comment (jemiluv8) [reply packages/features/ee/workflows/lib/reminders/templates/whatsapp/whatsappEventCancelledTemplate.ts]: resolved

File: packages/features/ee/workflows/pages/workflow.tsx
Comment (jemiluv8) [reply packages/features/ee/workflows/pages/workflow.tsx]: resolved.

File: packages/features/ee/workflows/components/WorkflowDetailsPage.tsx
Comment (jemiluv8) [reply packages/features/ee/workflows/components/WorkflowDetailsPage.tsx]: Reverted to original. Resolved.

File: packages/features/ee/workflows/components/AddActionDialog.tsx
Comment (jemiluv8) [reply packages/features/ee/workflows/components/AddActionDialog.tsx]: resolved

File: packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql
Comment (jemiluv8) [reply packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql:6]: I'll do that.

File: packages/features/ee/workflows/pages/workflow.tsx
Comment (jemiluv8) [reply packages/features/ee/workflows/pages/workflow.tsx:145]: resolved

File: apps/web/public/static/locales/en/common.json
Comment (jemiluv8) [reply apps/web/public/static/locales/en/common.json:1484]: Not sure about this.

There can be SMS and WHATSAPP actions associated with the same event type so
the client may end up receiving both SMS or WHATSAPP actions (for the same event)
depending on the workflows associated with the event type. 

This means it may not just be a case of SMS or Whatsapp it could be
1. SMS
2. Whatsapp
3. SMS and WHATSAPP

Comment (jemiluv8) [reply apps/web/public/static/locales/en/common.json:1484]: Kindly check [this](https://github.com/calcom/cal.com/blob/931763959b288776514e0b90597fe18d2dc7cf08/packages/features/bookings/lib/getBookingFields.ts#L22) and related code for context and let me know how we proceed.

Comment (CarinaWolli) [reply apps/web/public/static/locales/en/common.json:1484]: I think to keep it simple we can just call it "Text notifications" instead

Comment (jemiluv8) [reply apps/web/public/static/locales/en/common.json]: I've updated this to show "Phone Number (Text Notifications)".
Changed the related locale variable to number_text_notifications

File: .env.example
Comment (jemiluv8) [reply .env.example]: removed

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Found some more issues: 
1) I saved a phone number into my workflow, when I revisit the workflow I don't see the number anymore but it says still verified. This looks like a new issue to me:
Screenshot 2023-07-03 at 15 56 22

2) I am able to choose WhatsApp templates for non-WhatsApp actions: 
- choose a WhatsApp action 
- choose completed template
- change action to email action 
- see that template is still completed (we should fall back to custom)
- save workflow with the completed template

Again a big thank you for your work so far, really love that PR 🙏 Just tested the reminder and completed template outside of sandbox mode and everything worked 🙌🏻 (We maybe have to disable the other two templates for now until I get them approved in Twilio)

**Comment by CarinaWolli** [comment]
> I've tried the same test in chrome and firefox. I was able to replicate it in chrome but not firefox.
> So it works as expected in firefox. Can you confirm this?

Yes can confirm, it works in Firefox not in Chrome

> Should I go ahead and disable the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

Yes, let's disable it for now. We should still keep all the migrations and everything, let's just not show them in the template dropdown. We want it to be simple to enable them once Twilo approves them

**Comment by jemiluv8** [comment]
I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

**Comment by CarinaWolli** [comment]
> I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

@jemiluv8  Perfect! Any ideas already why the phone number isn't showing in Chrome?

File: packages/features/ee/workflows/lib/constants.ts
Comment (CarinaWolli) [inline packages/features/ee/workflows/lib/constants.ts:26]:
```suggestion
```

**Comment by CarinaWolli** [comment]
@jemiluv8 Good news, the templates just got approved 🎉

I will test them in a little bit, but we can revert the changes now :)

File: packages/features/ee/workflows/lib/constants.ts
Comment (jemiluv8) [reply packages/features/ee/workflows/lib/constants.ts:26]: I've reverted the changes to remove the "CANCELLED" and "RESCHEDULED" templates

Comment (CarinaWolli) [reply packages/features/ee/workflows/lib/constants.ts:26]: Can you remove the duplicate `WorkflowTemplates.REMINDER` in `WORKFLOW_TEMPLATES` as well please

**Comment by jemiluv8** [comment]
> > I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.
> 
> @jemiluv8 Perfect! Any ideas already why the phone number isn't showing in Chrome?

@CarinaWolli 
At the moment no. But I'm of the view that it is a general issue that may have to do with the
react-phone-input-2 introduced not long ago. I'd recommend we wait for a fix before "potentially"
merging this PR.

File: packages/features/ee/workflows/lib/constants.ts
Comment (jemiluv8) [reply packages/features/ee/workflows/lib/constants.ts]: I just did

**Comment by jemiluv8** [comment]
@CarinaWolli I'm still looking into the phone input not showing in chrome.
During my experiments I've noticed that I sometimes got the number to 
show while editing code (after hot reload).

Sometimes the number shoes and disappears - very fast. I should be able to get
a fix soon.

**Comment by jemiluv8** [comment]
> @jemiluv8 Let me know in case you need help for the phone number issue, I think that's the last thing missing now 🙏

@CarinaWolli I do need help. But perhaps starting tomorrow if I'm not able to come up with a solution.
I've been trying stuff without a lot of progress but I haven't yet exhausted all my possible attempts at a solution.

**Comment by jemiluv8** [comment]
@CarinaWolli, I just pushed a hot fix for the phone number thing.
Can you confirm it works in chrome as well.

**Review by CarinaWolli — APPROVED** [review]
I made some minor final adjustments, and now we are good to go 🙌🏻
Thank you for your excellent work 🙏

**Pull request merged**
